### PR TITLE
Move to "replica" terminology, in line with redis v5+

### LINKED ISSUE #945: Track and reflect nomenclature changes re replication (master/slave)
Potential server-side changes re terminology of master/slave.

As those changes become more concrete:

- client lib MUST: be aware of (i.e. work correctly with) any server-side changes to commands and return values, including any version-specific semantics
- client lib SHOULD: use preferred nomenclature where possible - happy to add new methods/enums etc and add `[Obsolete(..., false)]` to the old
- documentation SHOULD: use preferred nomenclature throughout (perhaps with "replica (historically: 'slave')" clarifications)

see https://github.com/antirez/redis/issues/5335

open question: does this impact the "no more 1.* builds" decision?

### PR DESCRIPTION
**Pull request by mgravell**
Warning: to anyone jumping in: please restrict discussion here to genuine technical issues. It happening isn't up for debate.

Fixes #945

---

Historically, we stated that we would reflect redis terminology; in redis 5, the replication terminology was updated to master/replica, so: let's reflect that in the API

- new API methods / properties / enums added
- issue the new commands / keywords when we believe we are on a 5+ server
- recognize old *and* new replies
- old marked `[Obsolete]` with guidance, and are proxies to the new
- documentation/comments updated
- tests and test configs updated

Todo:

- [x] fix NRediSearch tests on CI server
- [x] add tests for enum string formatting of new aliases

### CODE REVIEW

File: src/StackExchange.Redis/CommandMap.cs [hunk 1]
```diff
@@ -49,8 +49,8 @@ internal CommandMap(CommandBytes[] map)
             RedisCommand.ECHO, RedisCommand.PING, RedisCommand.QUIT, RedisCommand.SELECT,
 
             RedisCommand.BGREWRITEAOF, RedisCommand.BGSAVE, RedisCommand.CLIENT, RedisCommand.CLUSTER, RedisCommand.CONFIG, RedisCommand.DBSIZE,
-            RedisCommand.DEBUG, RedisCommand.FLUSHALL, RedisCommand.FLUSHDB, RedisCommand.INFO, RedisCommand.LASTSAVE, RedisCommand.MONITOR, RedisCommand.SAVE,
-            RedisCommand.SHUTDOWN, RedisCommand.SLAVEOF, RedisCommand.SLOWLOG, RedisCommand.SYNC, RedisCommand.TIME
+            RedisCommand.DEBUG, RedisCommand.FLUSHALL, RedisCommand.FLUSHDB, RedisCommand.INFO, RedisCommand.LASTSAVE, RedisCommand.MONITOR, RedisCommand.REPLICAOF,
+            RedisCommand.SAVE, RedisCommand.SHUTDOWN, RedisCommand.SLAVEOF, RedisCommand.SLOWLOG, RedisCommand.SYNC, RedisCommand.TIME
```
Comment (NickCraver) [inline src/StackExchange.Redis/CommandMap.cs:53]: Not for PR, but let's put these all on their own line for easier edits later (ones and group only related ones like above)

File: src/StackExchange.Redis/ConnectionMultiplexer.cs [hunk 2]
```diff
@@ -429,19 +430,19 @@ internal void MakeMaster(ServerEndPoint server, ReplicationChangeOptions options
                 }
             }
 
-            // deslave...
+            // de-replicate
```
Comment (NickCraver) [inline src/StackExchange.Redis/ConnectionMultiplexer.cs:433]:
```suggestion
            // stop replicating, promote to a standalone primary
```

File: src/StackExchange.Redis/ConnectionMultiplexer.cs [hunk 3]
```diff
@@ -480,17 +481,17 @@ void Broadcast(ReadOnlySpan<ServerEndPoint> serverNodes)
                 }
             }
 
-            // Send a message before it happens - because afterwards a new slave may be unresponsive
+            // Send a message before it happens - because afterwards a new replica may be unresponsive
             Broadcast(nodes);
 
-            if ((options & ReplicationChangeOptions.EnslaveSubordinates) != 0)
+            if ((options & ReplicationChangeOptions.ReplicateToSubordinates) != 0)
```
Comment (NickCraver) [inline src/StackExchange.Redis/ConnectionMultiplexer.cs:487]: This one seems odd (but works) to me, the action isn't *from the master* as the verbage hints it, it's an action *on the subordinates*. Trying to think of a better name while reading this...

File: src/StackExchange.Redis/Enums/ClientFlags.cs [hunk 4]
```diff
@@ -59,5 +69,17 @@ public enum ClientFlags : long
...
+        /// </summary>
+        PubSubSubscriber = 512,
+        /// <summary>
+        /// the client is in readonly mode against a cluster node
+        /// </summary>
+        ReadOnlyCluster = 512,
+        /// <summary>
+        /// the client is connected via a Unix domain socket
+        /// </summary>
+        UnixDomainSocket = 512,
```
Comment (NickCraver) [inline src/StackExchange.Redis/Enums/ClientFlags.cs:83]: Are these intended additions in this PR? Doesn't seem related...and they're all 512

File: src/StackExchange.Redis/Enums/ClientType.cs [hunk 5]
```diff
@@ -8,14 +10,19 @@ public enum ClientType
...
         /// <summary>
         /// Replication connections
         /// </summary>
-        Slave,
+        [Obsolete(Messages.PreferReplica)]
+        Slave = 1,
+        /// <summary>
+        /// Replication connections
+        /// </summary>
+        Replica = 1,
```
Comment (NickCraver) [inline src/StackExchange.Redis/Enums/ClientType.cs:22]:
```suggestion
        Replica = 1, // as an implementation detail, note that enum.ToString prefers *later* options when naming Flags
```

File: src/StackExchange.Redis/Enums/ReplicationChangeOptions.cs [hunk 6]
```diff
@@ -21,12 +21,17 @@ public enum ReplicationChangeOptions
...
         /// <summary>
-        /// Issue a SLAVEOF to all other known nodes, making this this master of all
+        /// Issue a REPLICAOF to all other known nodes, making this this master of all
         /// </summary>
+        [Obsolete(Messages.PreferReplica)]
         EnslaveSubordinates = 4,
         /// <summary>
+        /// Issue a REPLICAOF to all other known nodes, making this this master of all
+        /// </summary>
+        ReplicateToSubordinates = 4, // note ToString prefers *later* options
```
Comment (NickCraver) [inline src/StackExchange.Redis/Enums/ReplicationChangeOptions.cs:31]: `FormReplicaChain`? I don't know that I have an actual better suggestion here - whatever you think

File: src/StackExchange.Redis/Interfaces/IServer.cs [hunk 7]
```diff
@@ -33,15 +33,27 @@ public partial interface IServer : IRedis
...
+        /// <summary>
+        /// Explicitly opt in for replica writes on writable replica
+        /// </summary>
+        [Obsolete(Messages.PreferReplica)]
         bool AllowSlaveWrites { get; set; }
 
+        /// <summary>
+        /// Explicitly opt in for replica writes on writable replica
+        /// </summary>
+        bool AllowReplicaWrites { get; set; }
```
Comment (NickCraver) [inline src/StackExchange.Redis/Interfaces/IServer.cs:55]: How concerned are we about the interface break here?

File: src/StackExchange.Redis/Interfaces/IServer.cs [hunk 8]
```diff
@@ -33,15 +33,27 @@ public partial interface IServer : IRedis
...
+        /// Gets whether the connected server is a replica
         /// </summary>
+        [Obsolete(Messages.PreferReplica)]
         bool IsSlave { get; }
 
         /// <summary>
-        /// Explicitly opt in for slave writes on writable slaves
+        /// Gets whether the connected server is a replica
         /// </summary>
+        bool IsReplica { get; }
```
Comment (NickCraver) [inline src/StackExchange.Redis/Interfaces/IServer.cs:44]: Same as below

File: src/StackExchange.Redis/RedisServer.cs [hunk 9]
```diff
@@ -638,12 +646,14 @@ public void SlaveOf(EndPoint master, CommandFlags flags = CommandFlags.None)
             }
         }
 
-        public Task SlaveOfAsync(EndPoint master, CommandFlags flags = CommandFlags.None)
+        Task IServer.SlaveOfAsync(EndPoint master, CommandFlags flags) => ReplicaOfAsync(master, flags);
```
Comment (NickCraver) [inline src/StackExchange.Redis/RedisServer.cs:649]: This went private - public API break when we should obsolete?

File: src/StackExchange.Redis/RedisServer.cs [hunk 10]
```diff
@@ -856,14 +866,22 @@ public Task SentinelFailoverAsync(string serviceName, CommandFlags flags = Comma
             return ExecuteAsync(msg, ResultProcessor.SentinelArrayOfArrays);
         }
 
-        public KeyValuePair<string, string>[][] SentinelSlaves(string serviceName, CommandFlags flags = CommandFlags.None)
+        KeyValuePair<string, string>[][] IServer.SentinelSlaves(string serviceName, CommandFlags flags)
```
Comment (NickCraver) [inline src/StackExchange.Redis/RedisServer.cs:869]: API break? It's a recent addition, but still

File: src/StackExchange.Redis/RedisServer.cs [hunk 11]
```diff
@@ -856,14 +866,22 @@ public Task SentinelFailoverAsync(string serviceName, CommandFlags flags = Comma
...
+
+        public KeyValuePair<string, string>[][] SentinelReplicas(string serviceName, CommandFlags flags = CommandFlags.None)
         {
+            // note: sentinel does not have "replicas" terminology at the current time
             var msg = Message.Create(-1, flags, RedisCommand.SENTINEL, RedisLiterals.SLAVES, (RedisValue)serviceName);
             return ExecuteSync(msg, ResultProcessor.SentinelArrayOfArrays);
         }
 
-        public Task<KeyValuePair<string, string>[][]> SentinelSlavesAsync(string serviceName, CommandFlags flags = CommandFlags.None)
+        Task<KeyValuePair<string, string>[][]> IServer.SentinelSlavesAsync(string serviceName, CommandFlags flags)
```
Comment (NickCraver) [inline src/StackExchange.Redis/RedisServer.cs:879]: API break? It's a recent addition, but still

File: src/StackExchange.Redis/ResultProcessor.cs [hunk 12]
```diff
@@ -758,17 +759,17 @@ protected override bool SetResultCore(PhysicalConnection connection, Message mes
...
-                                        server.Multiplexer.Trace("Auto-configured slave-read-only: true");
+                                        server.ReplicaReadOnly = true;
+                                        server.Multiplexer.Trace("Auto-configured replica-read-only: true");
                                     }
                                     else if (val.IsEqual(CommonReplies.no))
                                     {
-                                        server.SlaveReadOnly = false;
-                                        server.Multiplexer.Trace("Auto-configured slave-read-only: false");
+                                        server.ReplicaReadOnly = false;
+                                        server.Multiplexer.Trace("Auto-configured replica-read-only: false");
```
Comment (NickCraver) [inline src/StackExchange.Redis/ResultProcessor.cs:772]: I'm good with this, just noting it's not technically accurate as we could be configuring a pre-5 server and this isn't the config value - don't care, log says what we need?

File: src/StackExchange.Redis/ServerEndPoint.cs [hunk 13]
```diff
@@ -642,7 +642,7 @@ internal Task<bool> SendTracer(LogProxy log = null)
         internal string Summary()
         {
             var sb = new StringBuilder(Format.ToString(EndPoint))
-                .Append(": ").Append(serverType).Append(" v").Append(version).Append(", ").Append(isSlave ? "slave" : "master");
+                .Append(": ").Append(serverType).Append(" v").Append(version).Append(", ").Append(isReplica ? "slave" : "master");
```
Comment (NickCraver) [inline src/StackExchange.Redis/ServerEndPoint.cs:645]:
```suggestion
                .Append(": ").Append(serverType).Append(" v").Append(version).Append(", ").Append(isReplica ? "replica" : "master");
```

File: src/StackExchange.Redis/StringSplits.cs [hunk 14]
```diff
@@ -6,4 +6,9 @@ internal static class StringSplits
             Space = { ' ' },
             Comma = { ',' };
     }
+
+    internal static class Messages
+    {
+        public const string PreferReplica = "Starting with Redis version 5, redis has moved to 'replica' terminology.";
```
Comment (NickCraver) [inline src/StackExchange.Redis/StringSplits.cs:12]: I think this is fine for most places, but probably a better "use this property" instead directly pointing to the replacement would be better - I'm happy to do that commit if cool with it. Related: a "Messages" internal class name here probably isn't awesome, should be `ObsoleteMessages` or something since Message is a core thing.

Cool with me putting specific messages and removing this, bypassing the name issue?

**Comment by mgravell** [comment]
New enums were new options I found while checking no "R" for replica.
Added, all 512 was error.

Property break on IServer is error.

Will fix.

File: src/StackExchange.Redis/Interfaces/IServer.cs (hunk 7 above)
Comment (NickCraver) [reply src/StackExchange.Redis/Interfaces/IServer.cs:55]: Note: we break the interface _no matter what_ here - so what's the thinking on best approach? I'm absolutely willing to take a major version here, but it's a nontrivial break regardless - not sure what's the best course of action

File: src/StackExchange.Redis/StringSplits.cs (hunk 14 above)
Comment (mgravell) [reply src/StackExchange.Redis/StringSplits.cs]: How about as written plus suffix " please use " + nameof

File: src/StackExchange.Redis/RedisServer.cs (hunk 11 above)
Comment (mgravell) [reply src/StackExchange.Redis/RedisServer.cs:879]: Thought I forwarded. Should have.

File: src/StackExchange.Redis/StringSplits.cs (hunk 14 above)
Comment (NickCraver) [reply src/StackExchange.Redis/StringSplits.cs]: Agree that sounds good - will adjust

File: src/StackExchange.Redis/RedisServer.cs (hunk 11 above)
Comment (NickCraver) [reply src/StackExchange.Redis/RedisServer.cs:879]: Ah yes this one's internal :+1: All good

**Comment by NickCraver** [comment]
@mgravell If you want to handle API bits...I'm trying to figure out why search CI is failing, locally I'm getting a log full of things like:
```
redis_1  | 50:S 07 Jun 2020 15:44:02.451 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.SAFEADD`, with args beginning with: `ClientTests/ClientTest.cs:TestHighlightSummarize`, `foo`, `1`, `FIELDS`, `text`, `Redis is often referred as a data structures s`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.467 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.DROP`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.469 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.CREATE`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, `SCHEMA`, `title`, `TEXT`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.471 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.SAFEADD`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, `doc0`, `1`, `FIELDS`, `title`, `hello world`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.472 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.SAFEADD`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, `doc1`, `1`, `FIELDS`, `title`, `hello world`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.474 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.SAFEADD`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, `doc2`, `1`, `FIELDS`, `title`, `hello world`, ' after processing the command '<unknown>'
redis_1  | 50:S 07 Jun 2020 15:44:02.475 # == CRITICAL == This replica is sending an error to its master: 'unknown command `FT.SAFEADD`, with args beginning with: `ClientTests/ClientTest.cs:TestDrop`, `doc3`, `1`, `FIELDS`, `title`, `hello world`, ' after processing the command '<unknown>'
```
...trying to figure it out now.

File: src/StackExchange.Redis/Interfaces/IServer.cs (hunk 7 above)
Comment (mgravell) [reply src/StackExchange.Redis/Interfaces/IServer.cs:55]: if by "interface break" you mean "adding things", I'm not in the least bit concerned by that

Comment (NickCraver) [reply src/StackExchange.Redis/Interfaces/IServer.cs:58]: Yep, exactly that - good with me :+1:

**Comment by NickCraver** [comment]
Ah, I see - 2 errors here:
1. The module is failing to load on the master - looks like it's on latest Redis and our module version breaks:
```
redis_1  | 52:M 07 Jun 2020 15:43:30.860 * <ft> Initialized thread pool!
redis_1  | 52:M 07 Jun 2020 15:43:30.860 * Module 'ft' loaded from /data/Basic/redisearch.so
redis_1  | 52:M 07 Jun 2020 15:43:30.860 # Module /data/Basic/redisearch.so initialization failed. Module not loaded
redis_1  | 52:M 07 Jun 2020 15:43:30.860 # Module /data/Basic/redisearch.so initialization failed. Module not loaded
redis_1  | 52:M 07 Jun 2020 15:43:30.860 # Module /data/Basic/redisearch.so initialization failed. Module not loaded
redis_1  | 52:M 07 Jun 2020 15:43:30.860 # Module /data/Basic/redisearch.so initialization failed. Module not loaded
```
2. The module isn't loaded on the replica, that's existing fail so it doesn't understand the command.

I'll take a look at version pinning here.

**Comment by NickCraver** [comment]
@mgravell Okay this RediSearch thing isn't simple. I think what happened is: we have a version of the module for a version of Redis. The version of Redis on AppeVeyor for Ubuntu for bumped to 5.0.7 and it appears to be incompatible with our module version. It's not related to this PR at all since master branch was breaking just before as well.

Proposal:
I recommend we don't care about it here, just make sure all the other tests pass. I'll try to make NRediSearch use _their_ docker image and put it in docker compose, also switching the Ubuntu build to Docker usage. Locally, you'd need to be using the `docker-compose up` testing method to get NRediSearch tests, otherwise they'll skip as inconclusive. I'm thinking of trying to put it on :6385 (currently free), that also means we won't spam the master/replica log...since we're not really testing that anyway, :6379 was jut convenient when we set it up.

Thoughts? I can try to tackle that in another branch...docker-compose is fighting me a bit on the supervisor and I'm not sure why yet. I don't have any reason to believe anything is actually broken here (all green local and manually on latest)...it's purely a CI testing issue.

File: src/StackExchange.Redis/ConnectionMultiplexer.cs (hunk 3 above)
Comment (NickCraver) [reply src/StackExchange.Redis/ConnectionMultiplexer.cs:487]: Maybe `ReplicateToOtherEndpoints`? That's really what's happening here.

File: src/StackExchange.Redis/RedisFeatures.cs [hunk 15]
```diff
@@ -211,6 +220,11 @@ public RedisFeatures(Version version)
         /// </summary>
         public bool KeyTouch => Version >= v3_2_1;
 
+        /// <summary>
+        /// Does the server prefer 'replica' terminology?
+        /// </summary>
+        public bool PreferReplica => Version >= v5_0_0;
```
Comment (NickCraver) [inline src/StackExchange.Redis/RedisFeatures.cs:226]: Maybe `ReplicaCommands`? Since PreferReplica is another thing on command flags, just thinking out loud here.

File: src/StackExchange.Redis/RedisServer.cs (hunk 9 above)
Comment (mgravell) [reply src/StackExchange.Redis/RedisServer.cs:649]: `internal sealed class RedisServer : RedisBase, IServer` - it is accessed via the `IServer` interface; all this did was make the internal-code-accessible version the current one, and make the legacy one explicit interface implementation

File: src/StackExchange.Redis/ConnectionMultiplexer.cs (hunk 3 above)
Comment (mgravell) [reply src/StackExchange.Redis/ConnectionMultiplexer.cs]: done; also added format test

**Comment by mgravell** [comment]
catch up when you're around; I *think* I've resolved the outstanding questions

File: src/StackExchange.Redis/ServerEndPoint.cs [hunk 16]
```diff
@@ -312,9 +312,9 @@ internal void AutoConfigure(PhysicalConnection connection)
             }
             else if (commandMap.IsAvailable(RedisCommand.SET))
             {
-                // this is a nasty way to find if we are a slave, and it will only work on up-level servers, but...
+                // this is a nasty way to find if we are a replica, and it will only work on up-level servers, but...
                 RedisKey key = Multiplexer.UniqueId;
-                msg = Message.Create(0, flags, RedisCommand.SET, key, RedisLiterals.slave_read_only, RedisLiterals.PX, 1, RedisLiterals.NX);
+                msg = Message.Create(0, flags, RedisCommand.SET, key, RedisLiterals.replica_read_only, RedisLiterals.PX, 1, RedisLiterals.NX);
```
Comment (NickCraver) [inline src/StackExchange.Redis/ServerEndPoint.cs:317]: I think this needs ternary love

Comment (NickCraver) [reply src/StackExchange.Redis/ServerEndPoint.cs:317]: Maybe something like a "RedisReplicaCheck" one-off here?
